lib/scripts.js

```diff
@@ -355,26 +355,36 @@ var scripts = {
   },
 
   /**
-   * Gets a stalled job by locking it and checking it is not already completed.
-   * Returns a "OK" if the job was locked and not in completed set.
+   * Iterates to queue and moves all jobs in the active queue that appear to be
+   * stalled back to the wait state to be reprocessed.
    */
-  getStalledJob: function(queue, job, token){
+  processStalledJobs: function(queue, grace){
     var script = [
-      'if redis.call("sismember", KEYS[1], ARGV[1]) == 0 then',
-      '  return redis.call("set", KEYS[2], ARGV[2], "PX", ARGV[3], "NX")',
+      'local activeJobs = redis.call("LRANGE", KEYS[1], 0, -1)',
+      'local stalled = {}',
+      'for _, job in ipairs(activeJobs) do',
+      '  local jobKey = ARGV[2] .. job',
+      '  if(redis.call("EXISTS", jobKey .. ":lock") == 0) then',
+      '    local jobTS = redis.call("HGET", jobKey, "timestamp")',
+      '    if(jobTS and jobTS < ARGV[1]) then',
```

Comment:
Yes, you're right. This was only a hack to cover the majority case: where jobs were just created and about to be processed by the worker that moved.